test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/MaxInputBufferLengthTests.cs

```diff
@@ -0,0 +1,173 @@
+        [InlineData(null, true, false)]
+        [InlineData(null, false, false)]
+        public async Task LargeUpload(int? maxInputBufferLength, bool sendContentLengthHeader, bool expectPause)
+        {
+            // Parameters
+            var data = new byte[10 * 1024 * 1024];
+            var bytesWrittenTimeout = TimeSpan.FromMilliseconds(100);
+            var maxSendSize = 4096;
+
+            // Initialize data with random bytes
+            (new Random()).NextBytes(data);
+
+            var startReadingRequestBody = new ManualResetEvent(false);
+            var clientFinishedSendingRequestBody = new ManualResetEvent(false);
+            var bytesWrittenEvent = new AutoResetEvent(false);
+
+            using (var host = StartWebHost(maxInputBufferLength, data, startReadingRequestBody, clientFinishedSendingRequestBody))
+            {
+                using (var socket = CreateSocketForHttpPost(host, sendContentLengthHeader ? data.Length : -1))
+                {
+                    var bytesWritten = 0;
+
+                    var sendTask = Task.Run(() =>
+                    {
+                        while (bytesWritten < data.Length)
+                        {
+                            var size = Math.Min(data.Length - bytesWritten, maxSendSize);
+                            bytesWritten += socket.Send(data, bytesWritten, size, SocketFlags.None);
+                            bytesWrittenEvent.Set();
+                        }
+
+                        Assert.Equal(data.Length, bytesWritten);
+                        socket.Shutdown(SocketShutdown.Send);
+                        clientFinishedSendingRequestBody.Set();
+                    });
+
+                    if (expectPause)
+                    {
+                        // Block until the send task has gone a while without writing bytes, which likely means
+                        // the server input buffer is full.
+                        while (bytesWrittenEvent.WaitOne(bytesWrittenTimeout)) { }
+
+                        // Verify the number of bytes written before the client was paused.
+                        // 
+                        // The minimum is (maxInputBufferLength - maxSendSize + 1), since if bytesWritten is
+                        // (maxInputBufferLength - maxSendSize) or smaller, the client should be able to
+                        // complete another send.
+                        // 
+                        // The maximum is harder to determine, since there can be OS-level buffers in both the client
+                        // and server, which allow the client to send more than maxInputBufferLength before getting
+                        // paused.  We assume the combined buffers are smaller than the difference between
+                        // data.Length and maxInputBufferLength.                          
+                        Assert.InRange(bytesWritten, maxInputBufferLength.Value - maxSendSize + 1, data.Length - 1);
```

Comment:
This assertion is failing on AppVeyor and Travis. It seems that `bytesWritten` is usually (but not always) 0. E.g. [On AppVeyor:](https://ci.appveyor.com/project/aspnet-ci/kestrelhttpserver/build/1.0.1754#L530)

    Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.MaxInputBufferLengthTests.LargeUpload(maxInputBufferLength: 16384, sendContentLengthHeader: True, expectPause: True) [FAIL] 
      Assert.InRange() Failure 
      Range:  (12289 - 10485759) 
      Actual: 0 
      Stack Trace: 
        MaxInputBufferLengthTests.cs(79,0): at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.MaxInputBufferLengthTests.<LargeUpload>d__0.MoveNext() 
        --- End of stack trace from previous location where exception was thrown --- 
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
           at System.Runtime.CompilerSer